README.md
# @luckydrq/tcp-net

编写稳定可靠的TCP长连接应用。

封装TCP连接的拆解包逻辑，支持`Length-Based`和`Line-Based`两种协议格式。

**注意，本模块只支持node 8.x以上。**

[![NPM version][npm-image]][npm-url]
[![build status][travis-image]][travis-url]
[![Test coverage][codecov-image]][codecov-url]
[![David deps][david-image]][david-url]
[![Known Vulnerabilities][snyk-image]][snyk-url]
[![NPM download][download-image]][download-url]
[![Gitter][gitter-image]][gitter-url]

[npm-image]: https://img.shields.io/npm/v/@luckydrq/tcp-net.svg?style=flat-square
[npm-url]: https://npmjs.org/package/@luckydrq/tcp-net
[travis-image]: https://img.shields.io/travis/luckydrq/tcp-net.svg?style=flat-square
[travis-url]: https://travis-ci.org/luckydrq/tcp-net
[codecov-image]: https://img.shields.io/codecov/c/github/luckydrq/tcp-net.svg?style=flat-square
[codecov-url]: https://codecov.io/gh/luckydrq/tcp-net
[david-image]: https://img.shields.io/david/luckydrq/tcp-net.svg?style=flat-square
[david-url]: https://david-dm.org/luckydrq/tcp-net
[snyk-image]: https://snyk.io/test/npm/@luckydrq/tcp-net/badge.svg?style=flat-square
[snyk-url]: https://snyk.io/test/npm/@luckydrq/tcp-net
[download-image]: https://img.shields.io/npm/dm/@luckydrq/tcp-net.svg?style=flat-square
[download-url]: https://npmjs.org/package/@luckydrq/tcp-net
[gitter-image]: https://img.shields.io/gitter/room/luckydrq/tcp-net.svg?style=flat-square
[gitter-url]: https://gitter.im/luckydrq/tcp-net

## Feature

- [x] 长连接
- [x] 连接保持（心跳检测）
- [x] 支持多请求多响应
- [x] 可扩展协议
- [x] 可扩展编解码器

## Protocol

我们对一个tcp包协议格式进行了如下约定：

```
|-----------------------------------------------|
| 1 byte  | 4 bytes  | 1 byte     | N bytes     |
|-----------------------------------------------|
| reqType | packetId | packetType | custom body |
|-----------------------------------------------|
```

- reqType: Number, 请求类型。0 - 请求 1 - 响应
- packetId: Number, 包id。
- packetType: Number, 包类型。0 - 普通包 1 - 心跳包
- custom body: 自定义包体。

其中，自定义包体就是用户发送的数据，它可以支持`Length-Based`和`Line-Based`两种子协议。

### Length-Based Protocol

默认内置。适用于用户定义的包长度是确定的场景。原理很简单，先用一个field标识包体的长度，紧接着就是相应长度的包体。field本身长度支持自定义设置，默认是4字节。

```
|------------------|---------------|
|      4 bytes     |    N bytes    |
|------------------|---------------|
| PACKET LENGTH(N) | HEADER | BODY |
|------------------|---------------|
```

### Line-Based Protocol

适用于用户定义的包长度不确定的场景。它用一种边界符来表示包的结束。最典型的例子，`http`协议格式就是`Line-Based`的协议，它的边界符（或者说分隔符）是`CRLF(\r\n)`。边界符支持自定义设置，默认是`CRLF`。

```
|------------------|-----------|
|      N bytes     | delimiter |
|------------------|-----------|
|    PACKET BODY   |    \r\n   |
|------------------|-----------|
```

## Install
```
$ npm i @luckydrq/tcp-net -S
```

## Examples

### Basic

client和server应该使用相同的协议，保持拆解包逻辑一致。

#### server
```js
const net = require('net');
const { Connection } = require('@luckydrq/tcp-net');

net.createServer(socket => {
  const conn = new Connection({
    socket,
    // protocol: new LengthBasedProtocol(),  default
  });

  // 发送一个packet
  conn.write('hello world');

  // 收到packet
  conn.on('packet', ({ packetId, body }) => {
    console.log(packetId); // 1
    console.log(JSON.parse(body)); // { name: 'xuezu' }
  });
}).listen();

```

#### client

```js
const { Connection } = require('@luckydrq/tcp-net');
const conn = new Connection({ host, port });

// 发送一个packet
conn.write({ name: 'xuezu' });

// 收到一个packet
conn.on('packet', ({ packetId, body }) => {
  console.log(packetId); // 1
  console.log(body); // hello world
});
```

### Heartbeat

内置了心跳检测，默认每隔5秒发起一次心跳检测，响应超时时间为3秒，超时3次则关闭连接。支持配置，在构造函数中传入：

```js
const conn = new Connection({
  heartbeatInterval: 5 * 1000, // 心跳间隔
  heartbeatTimeout: 3 * 1000, // 超时时间
  heartbeatLimit: 3, // 允许超时次数
});

conn.on('heartbeatTimeout', () => {
  console.log('连接已经关闭');
});
```

### Transcoder

通常需要对数据做编解码，本例子展示如何接入一个自定义编解码器。如果不提供，默认[只做基本的序列化](https://github.com/luckydrq/tcp-net/blob/master/lib/protocol.js)。

自定义实现一个transCoder，只需要实现`.encode`和`.decode`方法即可，也可以只实现其中之一，不过通常编解码逻辑都是对应的。

#### server

```js
const transCoder = {
  encode(buf) {
    // 调用hessian编码
    // 注意，需要返回Buffer对象
    return hessian.encode(buf);
  },

  decode(data) {
    // 可以返回任意类型的数据
    return hessian.decode(buf);
  },
};

net.createServer(socket => {
  const conn = new Connection({
    socket,
    protocol: new LengthBasedProtocol({ transCoder }), // 在协议中传入transCoder
  });

  // 收到packet
  conn.on('packet', ({ body }) => console.log(body));
}).listen();
```

#### client

```js
const conn = new Connection({
  host,
  port,
  protocol: new LengthBasedProtocol({ transCoder }),
});
// 发一个packet
conn.write({ name: 'xuezu' });
```

### [更多例子](https://github.com/luckydrq/tcp-net/blob/master/test/index.test.js)



History.md

1.2.2 / 2023-07-20
==================

  * fix: check socket writable before write (#5)

1.2.1 / 2023-01-19
==================

  * fix(types): IConnectionOptions should accept host and port

1.2.0 / 2023-01-19
==================

  * feat: support types for TypeScript.

1.1.0 / 2019-01-21
==================

  * feat: support options.connect timeout (#1)

1.0.1 / 2019-01-07
==================

  * deps: upgrade deps to date

1.0.0 / 2019-01-07
==================

  * chore: bump to 1.0.0


test/protocol.test.js
/**!
 * tcp-net
 *
 * Authors:
 *   luckydrq <drqzju@gmail.com> (http://github.com/luckydrq)
 */

'use strict';

/**
 * Module dependencies.
 */

const assert = require('assert');
const net = require('net');
const mm = require('mm');
const {
  Connection,
  LengthBasedProtocol,
} = require('..');
const Protocol = require('../lib/protocol');

describe('test/protocol.test.js', () => {
  beforeEach(mm.restore);

  it('should implement #write method', done => {
    const server = net.createServer().listen(0, '127.0.0.1', () => {
      const host = server.address().address;
      const port = server.address().port;
      const client = new Connection({
        host,
        port,
        protocol: new Protocol(),
      }).on('connect', () => {
        client.on('error', e => {
          assert.equal(e.message, 'should be implemented!');
          done();
        });
        client.write('data');
      });
    });
  });

  it('should implement #read method', done => {
    const protocol = new Protocol();
    const server = net.createServer(socket => {
      new Connection({
        socket,
        protocol,
      }).on('error', e => {
        assert.equal(e.name, 'ConnectionReadError');
        assert.equal(e.message, 'should be implemented!');
        done();
      });
    }).listen(0, '127.0.0.1', () => {
      const host = server.address().address;
      const port = server.address().port;
      const client = new Connection({
        host,
        port,
        protocol: new LengthBasedProtocol(),
      }).on('connect', () => {
        client.write('data');
      });
    });
  });
});


index.d.ts
/// <reference types="node" />

declare module '@luckydrq/tcp-net' {
  import { Socket } from 'net';
  import SdkBase from 'sdk-base';

  interface ITransCoder {
    encode: (data: Buffer | Record<string, any> | string | number | boolean | null) => Buffer;
    decode: (data: Buffer) => string;
  }

  interface IProtocolOptions {
    encoding?: string;
    transCoder?: ITransCoder;
  }

  interface ILengthBasedProtocolOptions extends IProtocolOptions {
    packetSize?: number;
  }

  interface ILineBasedProtocolOptions extends IProtocolOptions {
    packetDelimiter: string;
  }

  interface IPacket {
    reqType: number;
    packetId?: number;
    packetType: number;
    body?: string;
  }

  abstract class Protocol extends SdkBase {
    constructor(options: IProtocolOptions);
    pack: (data: Buffer, pkt: IPacket) => { id: number; packet: Buffer };
    parse: (data: Buffer) => Buffer | null;
    encode: (data: Buffer | Record<string, any> | string | number | boolean | null) => Buffer;
    decode: (data: Buffer) => string;
  }

  class LengthBasedProtocol extends Protocol {
    constructor(options?: ILengthBasedProtocolOptions);
    read: (data: Buffer) => void;
    write: (data: Buffer, pkt?: IPacket) => void;
  }

  class LineBasedProtocol extends Protocol {
    constructor(options: ILineBasedProtocolOptions);
    read: (data: Buffer) => void;
    write: (data: Buffer, pkt?: IPacket, encoded?: boolean) => void;
  }

  interface IConnectionOptions {
    socket?: Socket;
    host?: string;
    port?: number;
    protocol?: Protocol;
    idleTimeout?: number;
    finWaitTimeout?: number;
    heartbeatInterval?: number;
    heartbeatTimeout?: number;
    heartbeatTimeoutLimit?: number;
  }

  class Connection extends SdkBase {
    constructor(options: IConnectionOptions);
    get connected(): boolean;
    get localAddress(): string | undefined;
    get remoteAddress(): string | undefined;
    get remotePort(): number | undefined;
    read: (data: Buffer) => void;
    write: <T = Buffer | Record<string, any> | string | number | boolean | null>(
      data: T,
      pkt?: IPacket
    ) => void;
    end: () => void;
  }

  export { Connection, LengthBasedProtocol, LineBasedProtocol, IPacket };
}


lib/parser.js
/**!
 * tcp-net
 *
 * Authors:
 *   luckydrq <drqzju@gmail.com> (http://github.com/luckydrq)
 */

'use strict';

/**
 * Module dependencies.
 */

const ByteBuffer = require('byte');
const EventEmitter = require('events');
const debug = require('debug')('net:parser');

class Parser extends EventEmitter {
  constructor(options) {
    super();

    // [
    //   { name: 'packetId', type: 'UInt32', size: 4, desc: '包id' },
    //   { name: 'packetType, type: 'UInt8', size: 1, desc: '包类型' },
    //   { name: 'requestType', type: 'UInt8', size: 1, desc: '请求类型' }
    //   { name: 'foo', type: 'String', size: 1, desc: '请求类型' }
    // ]

    // copy passed in fields
    this.fields = [];
    for (const field of options.fields) {
      const f = Object.assign({}, field);
      this._resetField(f);
      this.fields.push(f);
    }
  }

  get currentField() {
    for (const field of this.fields) {
      if (!field.parsed) {
        return field;
      }
    }
    return null;
  }

  parse(chunk) {
    const field = this.currentField;

    if (!field) {
      return this._reset()
        .parse(chunk);
    }

    const { size, buf } = field;
    buf.put(chunk);

    const offset = buf.position();
    if (offset < size) {
      return null;
    }

    // 读取field值
    this._parseFieldValue(field);
    debug('parsed field, name: %s, value: %s', field.name, field.value);
    this.emit('field', field);

    const left = field.buf.copy(size);

    // 包解析完毕
    if (!this.currentField) {
      const packet = this.fields.reduce((obj, { name, value }) => {
        obj[name] = value;
        return obj;
      }, {});
      this.emit('packet', packet);
      debug('emit packet: %j', packet);
      this._reset();
      return left;
    }

    return this.parse(left);
  }

  _reset() {
    // 重置field状态
    for (const field of this.fields) {
      this._resetField(field);
    }
    return this;
  }

  _resetField(field) {
    field.value = null;
    field.buf = new ByteBuffer({ size: field.size });
    field.parsed = false;
  }

  _parseFieldValue(field) {
    const { buf, type } = field;
    const index = 0;
    const value = buf[`get${type}`].apply(buf, [ index ]);
    field.value = value;
    field.parsed = true;
    return value;
  }


}

module.exports = Parser;


lib/protocol.js
/**!
 * tcp-net
 *
 * Authors:
 *   luckydrq <drqzju@gmail.com> (http://github.com/luckydrq)
 */

'use strict';

/**
 * Module dependencies.
 */

const assert = require('assert');
const Base = require('sdk-base');
const is = require('is-type-of');
const ByteBuffer = require('byte');
const debug = require('debug')('net:protocol');
const Parser = require('./parser');

const {
  FIELDS,
  MAX_PACKET_ID,
} = require('./constants');

/**
 * base protocol
 *
 * |-----------------------------------------------|
 * | 1 byte  | 4 bytes  | 1 byte     | N bytes     |
 * |-----------------------------------------------|
 * | reqType | packetId | packetType | custom body |
 * |-----------------------------------------------|
 *
 */

class Protocol extends Base {
  constructor(options = {}) {
    super(options);

    this._opaque = 0;
    this._size = 0;
    this._packet = null;
    this._parsed = false;
    this._fields = FIELDS;
    this._encoding = options.encoding || 'utf8';
    this._transCoder = options.transCoder;
    this._parser = new Parser({ fields: this._fields });
    this._parser.on('packet', packet => {
      this.parsed = true;
      this.packet = packet;
    });
  }

  get size() {
    if (!this._size) {
      for (const field of this._fields) {
        this._size += field.size;
      }
    }
    return this._size;
  }

  get parsed() {
    return this._parsed;
  }

  set parsed(val) {
    this._parsed = val;
  }

  get packet() {
    return this._packet;
  }

  set packet(packet) {
    this._packet = packet;
  }

  _generatePacketId() {
    if (this._opaque === MAX_PACKET_ID) {
      this._opaque = 1;
    } else {
      ++this._opaque;
    }
    return this._opaque;
  }

  pack(buf, info = {}) {
    const { reqType = 0, packetType = 0 } = info;
    let { packetId } = info;

    if (reqType === 0) {
      // 如果是普通请求，则为他生成packetId
      packetId = this._generatePacketId();
    } else if (reqType === 1) {
      // 如果是响应请求，则packetId不能为空
      assert(is.number(packetId), 'packetId should be number when reqType = 1!');
    } else {
      throw new Error('invalid reqType, should be either 0 or 1!');
    }
    info.packetId = packetId;
    info.packetType = packetType;
    debug('send packet(#%s)', packetId);

    const headerBuf = new ByteBuffer({ size: this.size });
    for (const field of this._fields) {
      headerBuf[`put${field.type}`].apply(headerBuf, [ info[field.name] ]);
    }
    return {
      id: packetId,
      packet: Buffer.concat([ headerBuf.copy(), buf ]),
    };
  }

  parse(chunk) {
    // 已经解析过了
    if (this.parsed) {
      return chunk;
    }

    chunk = this._parser.parse(chunk);
    if (chunk) {
      const { packetId } = this.packet;
      debug('receive packet(#%s)', packetId);
      return chunk;
    }
    return null;
  }

  encode(data) {
    const transCoder = this._transCoder;
    if (transCoder && transCoder.encode) {
      data = transCoder.encode(data);
      assert(is.buffer(data), 'transCoder.encode() should return Buffer!');
      return data;
    }
    return this._encode(data);
  }

  decode(data) {
    const transCoder = this._transCoder;
    if (transCoder && transCoder.decode) {
      return transCoder.decode(data);
    }
    return this._decode(data);
  }

  read() {
    throw new Error('should be implemented!');
  }

  write() {
    throw new Error('should be implemented!');
  }

  _encode(data) {
    if (data == null) {
      return Buffer.alloc(0);
    }

    if (Buffer.isBuffer(data)) {
      return data;
    }

    const encoding = this._encoding;
    if (is.object(data)) {
      data = JSON.stringify(data);
    } else {
      data = String(data);
    }
    return Buffer.from(data, encoding);
  }

  _decode(data) {
    return data.toString(this._encoding);
  }
}

module.exports = Protocol;


lib/constants.js
/**!
 * tcp-net
 *
 * Authors:
 *   luckydrq <drqzju@gmail.com> (http://github.com/luckydrq)
 */

'use strict';

/**
 * Module dependencies.
 */

exports.FIELDS = [
  {
    name: 'reqType',
    type: 'UInt8',
    size: 1,
    desc: '请求类型：0 - 请求，1 - 响应',
  },

  {
    name: 'packetId',
    type: 'UInt32',
    size: 4,
    desc: '包id',
  },

  {
    name: 'packetType',
    type: 'UInt8',
    size: 1,
    desc: '包类型：0 - 普通请求 1 - 心跳请求',
  },
];

exports.DISCONNECT = 'disconnect';
exports.DISCONNECTING = 'disconnecting';
exports.CONNECT = 'connect';
exports.CONNECTING = 'connecting';

exports.PACKET_ID_LENGTH = 4;

exports.MAX_PACKET_ID = Math.pow(2, 32) - 1;

exports.IDLE_TIMEOUT = 30 * 1000;

exports.FINWAIT_TIMEOUT = 2 * 60 * 1000;

exports.HEARTBEAT_INTERVAL = 5 * 1000;

exports.HEARTBEAT_TIMEOUT = 3 * 1000;

exports.HEARTBEAT_TIMEOUT_LIMIT = 3;

exports.LENGTH_BASED_PACKET_LENGTH = 4;

exports.LINE_BASED_PACKET_DELIMITER = '\r\n';


lib/protocols/line_based_protocol.js
/**!
 * tcp-net
 *
 * Authors:
 *   luckydrq <drqzju@gmail.com> (http://github.com/luckydrq)
 */

'use strict';

/**
 * Module dependencies.
 */

const assert = require('assert');
const is = require('is-type-of');
const ByteBuffer = require('byte');
const debug = require('debug')('net:line_based');
const Protocol = require('../protocol');
const { LINE_BASED_PACKET_DELIMITER } = require('../constants');

/**
 * line-based protocol
 *
 * |------------------|-----------|
 * |      N bytes     | delimiter |
 * |------------------|-----------|
 * |       body       |    \r\n   |
 * |------------------|-----------|
 *
 */

class LineBasedProtocol extends Protocol {
  constructor(options = {}) {
    super(options);

    if (options.packetDelimiter) {
      assert(is.string(options.packetDelimiter, 'options.packetDelimiter should be string!'));
    }

    this._packetDelimiter = Buffer.from(options.packetDelimiter || LINE_BASED_PACKET_DELIMITER);
    this._buf = new ByteBuffer(0);
    this._writeBuf = new ByteBuffer(0);
  }

  _parsePacketBody(index) {
    const buf = this._buf.copy(0, index);
    return this.decode(buf);
  }

  _reset(index) {
    this.parsed = false;

    const start = index + this._packetDelimiter.length;
    const end = this._buf.position();
    const buf = this._buf.copy(start, end);

    // 清空内部buffer
    this._buf.clear();

    // 如果有剩余，继续解析
    if (buf.length > 0) {
      this.read(buf);
    }
  }

  read(chunk) {
    debug('receive chunk: %j', chunk);

    chunk = this.parse(chunk);
    if (!chunk) {
      return;
    }

    this._buf.put(chunk);

    const delimiter = this._packetDelimiter;
    // 转成Buffer
    const buf = this._buf.copy();
    const index = buf.indexOf(delimiter);

    if (index > -1) {
      // 已得到完整的包
      const body = this._parsePacketBody(index);
      this.packet.body = body;
      this.emit('packet', this.packet);

      // 丢弃已经parse过的部分
      this._reset(index);
    }
  }

  write(data, info = {}, encoded = false) {
    debug('write packet: %j, %j', data, info);

    const delimiter = this._packetDelimiter;
    const { packetType } = info;
    if (packetType === 1) {
      // 心跳包直接封包
      const { id, packet } = this.pack(Buffer.concat([ delimiter ]), info);
      this.emit('data', packet);
      return id;
    }

    const buf = this._writeBuf;
    let chunk;
    if (data != null) {
      if (!encoded) {
        chunk = this.encode(data);
      } else {
        chunk = data;
      }
    } else {
      chunk = Buffer.alloc(0);
    }

    // 判断用户输入是否带有delimiter
    const index = chunk.indexOf(delimiter);
    if (index > -1) {
      // 含有分隔符，则对当前数据块进行切割
      const start = 0;
      const end = index + delimiter.length;
      buf.put(chunk.slice(start, end));

      // 已经有完整的包，进行封包
      const { packet } = this.pack(buf.copy(), info);
      this.emit('data', packet);

      // 清理内部缓冲区，把多余的chunk存入
      buf.clear();

      // chunk已经encode过了，不需要再encode
      return this.write(chunk.slice(end), info, true);
    }
    buf.put(chunk);


    return null;
  }
}

module.exports = LineBasedProtocol;


lib/protocols/length_based_protocol.js
/**!
 * tcp-net
 *
 * Authors:
 *   luckydrq <drqzju@gmail.com> (http://github.com/luckydrq)
 */

'use strict';

/**
 * Module dependencies.
 */

const assert = require('assert');
const is = require('is-type-of');
const ByteBuffer = require('byte');
const debug = require('debug')('net:length_based');
const Protocol = require('../protocol');
const { LENGTH_BASED_PACKET_LENGTH } = require('../constants');

/**
 * length-based protocol
 *
 * |------------------|---------------|
 * |      4 bytes     |    N bytes    |
 * |------------------|---------------|
 * | packet length(N) |      body     |
 * |------------------|---------------|
 *
 */

class LengthBasedProtocol extends Protocol {
  constructor(options = {}) {
    super(options);

    if (options.packetLength) {
      assert(is.number(options.packetLength), 'options.packetLength should be number!');
      assert(options.packetLength >= 4, 'options.packetLength should be larger than 4');
    }

    this._packetLength = options.packetSize || LENGTH_BASED_PACKET_LENGTH;
    this._buf = new ByteBuffer();
  }

  _parsePacketBodyLength() {
    return this._buf.getUInt32(0);
  }

  _parsePacketBody(start, end) {
    const buf = this._buf.copy(start, end);
    return this.decode(buf);
  }

  _reset(length) {
    this.parsed = false;

    const start = length;
    const end = this._buf.position();
    const buf = this._buf.copy(start, end);

    // 清空内部buffer
    this._buf.clear();

    // 如果有剩余，继续解析
    if (buf.length > 0) {
      this.read(buf);
    }
  }

  read(chunk) {
    chunk = this.parse(chunk);
    if (!chunk) {
      return;
    }

    this._buf.put(chunk);

    const packetLength = this._packetLength;
    const offset = this._buf.position();
    if (offset < packetLength) {
      return;
    }

    const packetBodyLength = this._parsePacketBodyLength();
    if (offset - packetLength >= packetBodyLength) {
      // 已经收到完整的packet
      const body = this._parsePacketBody(packetLength, packetLength + packetBodyLength);
      this.packet.body = body;
      this.emit('packet', this.packet);

      // 丢弃已经parse过的部分
      this._reset(packetLength + packetBodyLength);
    }
  }

  write(data, info = {}) {
    debug('write packet: %j, %j', data, info);

    let bodyBuf = Buffer.alloc(0);
    const headerBuf = Buffer.alloc(this._packetLength);

    bodyBuf = data != null ? this.encode(data) : Buffer.alloc(0);
    headerBuf.writeUInt32BE(bodyBuf.length);
    const totalBuf = Buffer.concat([ headerBuf, bodyBuf ]);
    const { id, packet } = this.pack(totalBuf, info);
    this.emit('data', packet);
    return id;
  }
}

module.exports = LengthBasedProtocol;


lib/connection.js
/**!
 * tcp-net
 *
 * Authors:
 *   luckydrq <drqzju@gmail.com> (http://github.com/luckydrq)
 */

'use strict';

/**
 * Module dependencies.
 */

const assert = require('assert');
const net = require('net');
const Base = require('sdk-base');
// const is = require('is-type-of');
const awaitEvent = require('await-event');
const debug = require('debug')('net:connection');
const LengthBasedProtocol = require('./protocols/length_based_protocol');
const {
  DISCONNECT,
  DISCONNECTING,
  CONNECT,
  CONNECTING,
  IDLE_TIMEOUT,
  FINWAIT_TIMEOUT,
  HEARTBEAT_INTERVAL,
  HEARTBEAT_TIMEOUT,
  HEARTBEAT_TIMEOUT_LIMIT,
} = require('./constants');
const DEFAULT_OPTIONS = {
  idleTimeout: 30 * 1000,
  finWaitTimeout: 2 * 60 * 1000,
};

class Connection extends Base {
  constructor(options = {}) {
    options.initMethod = '_connect';
    super(options);

    options = Object.assign({}, DEFAULT_OPTIONS, options);
    this._options = options;
    this._isServer = false;
    this._state = DISCONNECT;
    this._socket = options.socket;
    this._protocol = options.protocol || new LengthBasedProtocol();
    this._idleTimeout = options.idleTimeout || IDLE_TIMEOUT;
    this._finWaitTimeout = options.finWaitTimeout || FINWAIT_TIMEOUT;
    this._heartbeatTimer = null;
    this._heartbeatInterval = options.heartbeatInterval || HEARTBEAT_INTERVAL;
    this._heartbeatTimeout = options.heartbeatTimeout || HEARTBEAT_TIMEOUT;
    this._heartbeatTimeoutLimit = options.heartbeatTimeoutLimit || HEARTBEAT_TIMEOUT_LIMIT;
    this._heartbeatTimeoutCount = 0;
    this._pendingTimers = [];
    this.await = awaitEvent;

    this.ready(() => {
      // 服务端启动心跳检测
      if (this._isServer) {
        this._setupHeartbeatTimer();
      }
    });
  }

  async _connect() {
    if (this._socket) {
      // server
      this._isServer = true;
      process.nextTick(() => this._socket.emit('connect'));
    } else {
      // client
      assert(this._options.host, '[Connection]: options.host required!');
      assert(this._options.port, '[Connection]: options.port required!');
    }

    const { host, port } = this._options;
    this._state = CONNECTING;
    this._socket = this._socket || net.createConnection({ host, port });
    this._socket.on('connect', this._onConnect.bind(this));
    this._socket.on('data', data => this.read(data));
    this._protocol.on('data', data => this._socket.writable && this._socket.write(data));
    this._protocol.on('packet', this._onPacket.bind(this));
    this._protocol.on('error', this._onError.bind(this));
    this._socket.on('error', this._onError.bind(this));
    this._socket.on('close', this._onClose.bind(this));
    this._socket.on('timeout', this._onTimeout.bind(this));

    // 如果有传入连接超时，则先把timeout置为connectTimeout
    if (this._options.connectTimeout) {
      this._socket.setTimeout(this._options.connectTimeout);
    } else {
      this._socket.setTimeout(this._idleTimeout);
    }

    return this.await('connect');
  }

  _setupHeartbeatTimer() {
    const interval = this._heartbeatInterval;
    const timeout = this._heartbeatTimeout;
    const limit = this._heartbeatTimeoutLimit;

    // 心跳超时，关闭连接
    this.on('heartbeatTimeout', () => this.end());

    // 启动心跳定时器
    this._heartbeatTimer = setInterval(() => {
      let ack = false;

      // 发送心跳请求
      const packetId = this.write(null, { reqType: 0, packetType: 1 });
      const onAck = createOnAck(this, packetId);
      this.on('heartbeatPacket', onAck);
      const timer = setTimeout(() => {
        this.removeListener('heartbeatPacket', onAck);
        clearTimeout(timer);

        if (!ack) {
          // 心跳超时次数超过上限，触发heartbeatTimeout事件
          if (++this._heartbeatTimeoutCount >= limit) {
            this.emit('heartbeatTimeout');
          }
        }
      }, timeout);

      function createOnAck(conn, id) {
        return function onAck({ reqType, packetId, packetType }) {
          if (reqType === 1 && packetType === 1 && packetId === id) {
            ack = true;
            conn.removeListener('heartbeatPacket', onAck);
          }
        };
      }
    }, interval);
  }

  get connected() {
    return this._state === CONNECT;
  }

  get localAddress() {
    return this._socket && this._socket.localAddress;
  }

  get localPort() {
    return this._socket && this._socket.localPort;
  }

  get remoteAddress() {
    if (this._socket) {
      const remoteAddress = this._socket.remoteAddress;
      if (remoteAddress) {
        return remoteAddress;
      }
    }
    return this._options.host;
  }

  get remotePort() {
    if (this._socket) {
      const remotePort = this._socket.remotePort;
      if (remotePort) {
        return remotePort;
      }
    }
    return this._options.port;
  }

  read(data) {
    try {
      this._protocol.read(data);
    } catch (e) {
      e.name = 'ConnectionReadError';
      this.emit('error', e);
    }
  }

  write(data, info) {
    try {
      return this._protocol.write(data, info);
    } catch (e) {
      e.name = 'ConnectionWriteError';
      this.emit('error', e);
    }
  }

  end() {
    if (this._socket && this._socket.writable) {
      this._state = DISCONNECTING;
      this._socket.end();

      // ensure socket close after shutdown
      // https://github.com/nodejs/node/issues/11572
      const socket = this._socket;
      const timer = setTimeout(() => {
        if (socket && !socket.destroyed) {
          this.emit('finWaitTimeout');
          socket.destroy();
          this._onClose();
        }
      }, this._finWaitTimeout);
      this._pendingTimers.push(timer);
    }
    return this;
  }

  async _onConnect() {
    debug('connected: %j', this._options);

    // for test：为了模拟connect timeout超时的情况
    if (this._options.connectCallback) {
      await this._options.connectCallback();
    }

    // 重置一下，如果之前传入connectTimeout，则重置掉。
    // 重复调用socket.setTimeout没有问题，node内部会移除先前的listener: https://github.com/nodejs/node/blob/master/lib/net.js#L411-L413
    this._socket.setTimeout(this._idleTimeout);
    this._state = CONNECT;
    this.emit('connect');
  }

  _onPacket(packet) {
    debug('receive packet: %j', packet);

    const { reqType, packetId, packetType } = packet;
    if (packetType === 1) {
      if (reqType === 0) {
        // 心跳请求，一般是服务端发起
        this.write(null, { reqType: 1, packetId, packetType });
      } else if (reqType === 1) {
        // 心跳响应，把timeout次数减1，一般是客户端发起
        if (this._heartbeatTimeoutCount > 0) {
          --this._heartbeatTimeoutCount;
        }
      }
      this.emit('heartbeatPacket', packet);
      return;
    }

    this.emit('packet', packet);
  }

  _onTimeout() {
    debug('connection timeout!');

    const err = new Error(`timeout when connecting ${this.remoteAddress}`);
    err.name = 'ConnectionTimeOut';

    // 如果超时了，关闭连接
    this._onError(err);
  }

  _onError(e) {
    this.emit('error', e);
    this.end();
  }

  _onClose() {
    debug('closed: %j', this._options);

    // 连接已经关闭
    this._state = DISCONNECT;
    this.emit('close');
    this.removeAllListeners();

    if (this._heartbeatTimer) {
      clearInterval(this._heartbeatTimer);
    }

    // remove pending timers
    for (const timer of this._pendingTimers) {
      clearTimeout(timer);
    }
  }
}

module.exports = Connection;
